Testing matrix

gforcada opened the pull request:
Part of #234 

So far:

- split the gigantic `tox.ini.j2` into more sizeable chunks
- `tox` environments' names are generated correctly
- `tox` picks the correct python version
- multiple constraints files are specified so that each environment is using the correct one
- you can override the specific testing matrix on `.meta.toml` (if we had not enough configuration options 😓 )
- generate the right GitHub configuration 🎉 with proper names 😆 
- documentation!
- GitLab equivalent changes to the GitHub ones

gforcada committed:
chore(tox): split and reduce duplication

Tox configuration was getting too big to handle in a single file.

While doing that, take the time to reduce the duplication between the
test sections, as the differences are only on the requirements and the
commands, but everything else is the same.

gforcada committed:
breaking(tox): Plone and Python matrix testing

This allows running any combination of Plone and Python versions that
are configured on the repository itself, or the defaults provided by
`plone.meta` itself.

gforcada committed:
breaking(tox): use the right constraints file

Otherwise, the testing matrix is pointless!

Add a new setting for the other tox environments that also need a
constraints file, but not multiple.

gforcada committed:
Add news entries

1letter commented:
first quick test give me an error:
```

$ tox -e init
init: install_deps> python -I -m pip install mxdev
init: commands[0]> mxdev -c mx.ini
###############################################################################
# Load configuration
Can not parse override: 
###############################################################################
# Read infiles
Read [r]: requirements.txt
Read [c]: constraints.txt
Read [c]: https://dist.plone.org/release/6.1-latest/constraints.txt
###############################################################################
# No sources configured!
###############################################################################
# Write outfiles
Write [c]: constraints-mxdev.txt
Write [r]: requirements-mxdev.txt
🎂 You are now ready for: pip install -r requirements-mxdev.txt
   (path to pip may vary dependent on your installation method)
init: commands[1]> echo 'Initial setup for mxdev'
init: failed with echo is not allowed, use allowlist_externals to allow it
  init: FAIL code 1 (2.57 seconds)
  evaluation failed :( (2.69 seconds)
```

1letter commented on src/plone/meta/default/tox-init.j2:
perhaps a statement is missing?
```
allowlist_externals =
    echo
```

Workaround:

add this to .meta.toml:
```
testenv_options = """
allowlist_externals =
    echo
"""
```

gforcada commented:
@1letter oh, sorry, it was not meant to be tested still, but indeed, there is some polishing needed. Thanks for the heads up. After the next meeting I will fix it

gforcada committed:
fix(tox): adjust init environment

As the default `[testenv]` is taken over for the generated environments,
we need to set the settings which were defined there in the environments
that need it.

gforcada committed:
feat(gha): run part of the matrix

Like in Jenkins, for each plone version, run the tests against the
python versions upper and lower bounds.

gforcada committed:
breaking(gha): do not use plone.meta test workflow

gforcada committed:
feat(docs): update documentation and add upgrade guide

gforcada committed:
breaking(gitlab): run a test matrix

stevepiercy commented on README.md:
I skipped review of this file, as I already reviewed it in https://github.com/plone/meta/pull/258. After that PR is merged, then I'll revisit this file.

stevepiercy commented on news/234.breaking:
```diff
@@ -0,0 +1 @@
+Allow to define multiple plone and python versions in tox @gforcada
```
```suggestion
Support combinations of multiple versions of Plone and Python in tox. @gforcada
```

gforcada replied:
Though it is not only `tox`, it's being used in GHA and GitLab as well, that's why I kept it generic

stevepiercy commented on news/234.documentation:
```diff
@@ -0,0 +1 @@
+Add an upgrade guide @gforcada
```
```suggestion
Add an upgrade guide. @gforcada
```

stevepiercy commented on src/plone/meta/config_package.py:
```diff
@@ -369,15 +380,70 @@ def tox(self):
+                single_constraints = f"-c {constraints[first_plone_version]}"
+                if len(test_matrix.keys()) != len(constraints.keys()):
+                    raise ValueError(
+                        "`constraints_files` and `test_matrix` need to provide the same plone versions."
```
```suggestion
                        "`constraints_files` and `test_matrix` need to provide the same Plone versions."
```

stevepiercy commented on src/plone/meta/config_package.py:
```diff
@@ -369,15 +380,70 @@ def tox(self):
+                if len(test_matrix.keys()) != len(constraints.keys()):
+                    raise ValueError(
+                        "`constraints_files` and `test_matrix` need to provide the same plone versions."
+                        f"They are providing {list(constraints.keys())} and {list(test_matrix.keys())} respectively."
```
```suggestion
                        f"They provide {list(constraints.keys())} and {list(test_matrix.keys())}, respectively."
```

stevepiercy commented on src/plone/meta/config_package.py:
```diff
@@ -450,27 +515,70 @@ def gha_workflows(self):
+                    image = custom_images.get(py_version)
+                if not image:
+                    raise ValueError(
+                        f"There is no docker image defined for Python {py_version}. "
```
```suggestion
                        f"There is no Docker image defined for Python {py_version}. "
```

stevepiercy commented on src/plone/meta/config_package.py:
```diff
@@ -450,27 +515,70 @@ def gha_workflows(self):
+                if not image:
+                    raise ValueError(
+                        f"There is no docker image defined for Python {py_version}. "
+                        "Either provide it on `custom_images` option or report an issue to `plone.meta`."
```
```suggestion
                        "Either provide it in the `custom_images` option or report an issue to `plone.meta`."
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+# Upgrade guide
+
+Here are listed the breaking changes between major versions.
```
Do you want to document only breaking changes, or also include feature additions between minor versions and how to use them? If you want to include the latter, then I'd append ` and feature additions between minor versions`.

```suggestion
This file describes the breaking changes between major versions.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+
+You can now define the combinations of Plone versions and Python versions to be tested.
+
+`plone.meta` provides a default (the same as used by Plone itself),
```
```suggestion
`plone.meta` provides a default combination (the same as used by Plone itself), but can be overridden in the file `.meta.toml`.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+You can now define the combinations of Plone versions and Python versions to be tested.
+
+`plone.meta` provides a default (the same as used by Plone itself),
+but can be overridden on `.meta.toml`.
```
```suggestion
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+
+### Constraints
+
+The `constraints_file` option on `.meta.toml`'s `[tox]` table got renamed to `constraints_files`
```
```suggestion
The `constraints_file` option in `.meta.toml`'s `[tox]` table was renamed to `constraints_files`, and the type of its value was changed from a string to a dictionary.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+
+This option continues to be optional.
+
+The dictionary keys needs to be plone versions
```
```suggestion
The dictionary keys must be Plone versions, and each key's value must be the constraints file for that Plone version.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+
+### GitHub Actions
+
+The `py_versions` option on `.meta.toml`'s `[github]` table is deprecated,
```
```suggestion
The `py_versions` option in `.meta.toml`'s `[github]` table is deprecated.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+### GitHub Actions
+
+The `py_versions` option on `.meta.toml`'s `[github]` table is deprecated,
+we use the new `test_matrix` option from `[tox]` table
```
```suggestion
Use the new `test_matrix` option from the `[tox]` table, as you can now run multiple Python versions from within `tox` itself.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+
+### GitHub variables
+
+The GitHub variables `TEST_OS_VERSIONS` and `TEST_PYTHON_VERSIONS` are deprecated
```
```suggestion
The GitHub variables `TEST_OS_VERSIONS` and `TEST_PYTHON_VERSIONS` are deprecated and no longer used.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+
+### GitLab images
+
+The `custom_image` option on `.meta.toml`'s `[gitlab]` table got renamed to `custom_images`
```
```suggestion
The `custom_image` option in `.meta.toml`'s `[gitlab]` table was renamed to `custom_images`, and its value type was changed from a string to a dictionary.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,71 @@
+
+This option continues to be optional.
+
+The dictionary keys needs to be python versions
```
```suggestion
The dictionary keys must be Python versions, and the values a Docker image for that Python version.
```

stevepiercy reviewed (changes requested):
Some suggestions and a question. I didn't bother to submit multiple suggestions for a single sentence or concept. Narrative docs should follow the one sentence per line guideline of Plone Documentation, and not break up concepts for semantic line breaks or PEP8 line length. Semantic line breaks are harmful, and documentation is not code.

mauritsvanrees commented:
@gforcada I run into an error trying this.

I ran this on a branch of `collective.faq` where I am working on porting it to Plone 6.  See https://github.com/collective/collective.faq/pull/26.  The current tests (without this new testing matrix support) [fail](https://github.com/collective/collective.faq/actions/runs/13455178894/job/37597516054?pr=26) while trying to run buildout on a config that does not exist. :-)  So the testing matrix support could help there.

But when I run `config-package` on it, I get a traceback:

```
$ ./venv/bin/config-package --branch current /Users/maurits/zest/irceline6/src/collective.faq
*** towncrier: If you want to use Towncrier, you have to create a 'news/' folder manually.

Traceback (most recent call last):
  File "/Users/maurits/community/plone-coredev/6.2/src/meta/./venv/bin/config-package", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/maurits/community/plone-coredev/6.2/src/meta/src/plone/meta/config_package.py", line 747, in main
    package.configure()
  File "/Users/maurits/community/plone-coredev/6.2/src/meta/src/plone/meta/config_package.py", line 722, in configure
    files = method()
            ^^^^^^^^
  File "/Users/maurits/community/plone-coredev/6.2/src/meta/src/plone/meta/config_package.py", line 388, in tox
    options.update(self._handle_constraints_files(options))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/maurits/community/plone-coredev/6.2/src/meta/src/plone/meta/config_package.py", line 400, in _handle_constraints_files
    plone_versions = list(test_matrix.keys())
                          ^^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'keys'
```

I did not yet edit `.meta.toml` to do anything with a test matrix yet.  In fact, I have now basically emptied it from my previous manual attempt.

mauritsvanrees commented:
@gforcada The problem is that if `test_matrix` has not been adapted in `.meta.toml`, the `options` dict will contain a default value `"test_matrix" = ""`.

I have this change locally and then it works:

```
             test_matrix = options.get("test_matrix", TOX_TEST_MATRIX)
+            if not isinstance(test_matrix, dict):
+                test_matrix = TOX_TEST_MATRIX
             plone_versions = list(test_matrix.keys())
```

I am not sure this is the best way, but it helps.  And all jobs on my PR are green, for example [this one](https://github.com/collective/collective.faq/actions/runs/13455546852/job/37598670359?pr=26).  So looking good!

mauritsvanrees commented:
Ah, but the job called "6.0 on py3.13" actually has [Products.CMFPlone 6.1.0](https://github.com/collective/collective.faq/actions/runs/13455546852/job/37598670066?pr=26#step:6:22).  The other jobs are correct.

tox runs a [`pip install`](https://github.com/collective/collective.faq/actions/runs/13455546852/job/37598670066?pr=26#step:6:19) with `-c/home/runner/work/collective.faq/collective.faq/.tox/py313-plone60/constraints.txt`

Locally, `.tox/py313-plone60/constraints.txt` only has one line: `zope.testrunner`.
Ah, but locally `.tox/py39-plone60/constraints.txt` also has only this one line.  There, just like on GitHub Actions, the py39 one is using Plone 6.0, and py313 is using 6.1.

Strange.  There is some magic in tox that is not yet obvious to me.

mauritsvanrees commented:
* Locally py39-plone60 has `plone.staticresources` 2.2.3, which should be 2.1.17.
* Locally py313-plone60 has `plone.staticresources` 2.2.2, which should be 2.2.0.

I don't know why those versions are higher than they should be.

gforcada commented:
@stevepiercy fixes all applied, hopefully this time is all good from your side 🤞🏾 

@mauritsvanrees the `AttributeError` should be fixed, very weird the constraints problems, let me double check as well

gforcada commented:
@mauritsvanrees 

Somehow the `.tox/ENV/constraints.txt` only has `zope.testrunner` and no the complete Plone 6.0/6.1 constraints file ❗ 

In `.tox/test/constraints.txt` though, the complete constraints file is there 🤔

stevepiercy commented on README.md:
```diff
@@ -166,27 +166,28 @@ _your own configuration lines_
+> See the `test_matrix` option in [`tox`](#toxini) configuration file.
+
+> [!TIP]
+> 🍀 `plone.meta` tries to be a bit more environmental friendly.
```
```suggestion
> 🍀 `plone.meta` tries to be a bit more environmentally friendly.
```

stevepiercy commented on README.md:
```diff
@@ -166,27 +166,28 @@ _your own configuration lines_
 
 > [!NOTE]
-> The variables `TEST_OS_VERSIONS` and `TEST_PYTHON_VERSIONS` variables need to exist, either at the GitHub organization level, or at the repository level.
+> See the `test_matrix` option in [`tox`](#toxini) configuration file.
```
This relative anchor does not exist.

gforcada replied:
It should, it points to the 

```
#### `tox.ini`
```

and from what I see, the dot gets removed. I will double check though

gforcada replied:
@stevepiercy at least on GitHub UI if you click in [this link](https://github.com/plone/meta/tree/testing-matrix?tab=readme-ov-file#toxini) it leads you to the `tox.ini` section, and as you can see it removes the dot on the anchor.

We are expecting users to see this on GitHub, right? 🤔 so far this is not being imported in docs.plone.org

stevepiercy replied:
Ah, OK, ignore this one. I confirmed GitHub strips `.` from anchors.

stevepiercy commented on README.md:
```diff
@@ -253,13 +243,19 @@ _your own configuration lines_
-Specify a custom Docker image, if the default does not fit your needs, in the `custom_image` key.
+Specify custom Docker images in the `custom_images` key.
+
+The dictionary keys needs to be python versions
```
```suggestion
The dictionary keys needs to be Python versions and the value a Docker image for that Python version.
```

stevepiercy commented on README.md:
```diff
@@ -253,13 +243,19 @@ _your own configuration lines_
+Specify custom Docker images in the `custom_images` key.
+
+The dictionary keys needs to be python versions
+and the value a docker image for that python version.
```
Semantic line breaks are evil. Please don't use them.

```suggestion
```

stevepiercy commented on README.md:
```diff
@@ -493,14 +498,18 @@ test_deps_additional = """
 
+> [!WARNING]
+> You need to specify the same Plone versions as on `test_matrix` option
+> (or the default provided by `plone.meta`)
```
```suggestion
```

stevepiercy commented on README.md:
```diff
@@ -493,14 +498,18 @@ test_deps_additional = """
 There are plenty of tools that can do that though.
 
+> [!WARNING]
+> You need to specify the same Plone versions as on `test_matrix` option
```
```suggestion
> You need to specify the same Plone versions as in the `test_matrix` option or the default provided by `plone.meta`.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,68 @@
+This will be used to generate the necessary `tox` environments and the GitHub Actions that will be run on each pull request.
+
+> [!TIP]
+> 🍀 `plone.meta` tries to be a bit more environmental friendly.
```
```suggestion
> 🍀 `plone.meta` tries to be a bit more environmentally friendly.
```

stevepiercy commented on UPGRADE.md:
```diff
@@ -0,0 +1,68 @@
+### GitHub Actions
+
+The `py_versions` option in `.meta.toml`'s `[github]` table is deprecated.
+We use the new `test_matrix` option from `[tox]` table as we now can run multiple python versions from within `tox` itself.
```
```suggestion
We use the new `test_matrix` option from `[tox]` table, as we now can run multiple Python versions from within `tox` itself.
```

gforcada commented:
I found the problem:

The `[testenv]` section in `tox.ini` had a typo:

```diff
-deps = {[test_runner]deps}
+deps = {[base]deps}
```

I'm fixing it right away

stevepiercy reviewed (changes requested):
A few grammar and proper noun tweaks, and a few semantic line break habits to break, and it's good to go.

stevepiercy reviewed (approved):
Docs LGTM. Needs a technical review from someone more knowledgeable. Thank you!

mauritsvanrees reviewed (approved):
Yes, with your last fix it works on my `collective.faq` PR. 
This is very helpful.  Many thanks!

davisagli commented on README.md:
````diff
@@ -449,6 +444,15 @@ _your own configuration lines_
+```toml
+[tox]
+test_matrix = { "6.1" = ["3.13", "3.10"], "6.0": ["3.13", "3.9"] }
+```
````
We should say a bit more explicitly how this works. It's not obvious which one is the Plone version and which one is the Python version unless you recognize the numbers.

gforcada replied:
Oh, I think I mention it on the upgrade docs but not on the readme indeed.

davisagli reviewed (commented):
@gforcada Thanks, it's really nice to have support for this!
